File: chrome/browser/icon_loader.cc
// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/icon_loader.h"

#include "base/bind.h"
#include "base/threading/thread_task_runner_handle.h"
#include "content/public/browser/browser_thread.h"

using content::BrowserThread;

IconLoader::IconLoader(const base::FilePath& file_path,
                       IconSize size,
                       Delegate* delegate)
    : file_path_(file_path),
      icon_size_(size),
      delegate_(delegate) {}

IconLoader::~IconLoader() {
}

void IconLoader::Start() {
  target_task_runner_ = base::ThreadTaskRunnerHandle::Get();

  BrowserThread::PostTaskAndReply(BrowserThread::FILE, FROM_HERE,
      base::Bind(&IconLoader::ReadGroup, this),
      base::Bind(&IconLoader::OnReadGroup, this));
}

void IconLoader::ReadGroup() {
  group_ = GroupForFilepath(file_path_);
}

void IconLoader::OnReadGroup() {
  BrowserThread::PostTask(ReadIconThreadID(), FROM_HERE,
                          base::Bind(&IconLoader::ReadIcon, this));
}

void IconLoader::NotifyDelegate() {
  delegate_->OnImageLoaded(this, std::move(image_), group_);
}

File: chrome/browser/resources/downloads/icon_loader.js
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

cr.define('downloads', function() {
  var IconLoader = {
    /** @private {boolean} */
    isIconLoading_: false,

    /** @private {Array<{img: HTMLImageElement, url: string}>} */
    iconsToLoad_: [],

    /**
     * Load the provided |url| into |img.src| after appending ?scale=.
     * @param {!HTMLImageElement} img An <img> to show the loaded image in.
     * @param {string} url A remote image URL to load.
     */
    loadScaledIcon: function(img, url) {
      var scale = '?scale=' + window.devicePixelRatio + 'x';
      this.iconsToLoad_.push({img: img, url: url + scale});
      this.loadNextIcon_();
    },

    /** @private */
    loadNextIcon_: function() {
      if (this.isIconLoading_)
        return;

      this.isIconLoading_ = true;

      while (this.iconsToLoad_.length) {
        var request = IconLoader.iconsToLoad_.shift();
        var img = request.img;

        if (img.src == request.url)
          continue;

        img.onabort = img.onerror = img.onload = function() {
          this.isIconLoading_ = false;
          this.loadNextIcon_();
        }.bind(this);

        img.src = request.url;
        return;
      }

      // If we reached here, there's no more work to do.
      this.isIconLoading_ = false;
    },
  };

  return {IconLoader: IconLoader};
});
